File: sweetfactory-compiler/src/main/java/sweetfactory/compiler/SweetFactoryProcessor.java
package sweetfactory.compiler;

import com.squareup.javapoet.*;
import sweetfactory.annotations.*;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.*;
import javax.lang.model.type.MirroredTypeException;
import javax.lang.model.type.TypeMirror;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.*;

public class SweetFactoryProcessor extends AbstractProcessor {

    private MessagerUtil messagerUtil;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnvironment) {
        super.init(processingEnvironment);

        messagerUtil = new MessagerUtil(processingEnvironment.getMessager());
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnvironment) {
        for (Element element : roundEnvironment.getElementsAnnotatedWith(SweetFactoryDeclaration.class)) {
            TypeElement typeElement = (TypeElement) element;

            JavaFile javaFile = generateJavaFile(typeElement);

            try {
                javaFile.writeTo(processingEnv.getFiler());
            } catch (IOException e) {
                final String className = javaFile.toJavaFileObject().getName().replace("/", ".");
                messagerUtil.error(element, "Error when trying to generate code for %s", className);
            }
        }

        return false;
    }

    private JavaFile generateJavaFile(TypeElement typeElement) {
        Annotation intentFactoryAnnotation = typeElement.getAnnotation(SweetFactoryDeclaration.class);

        TypeMirror typeMirror = null;
        try {
            ((SweetFactoryDeclaration) intentFactoryAnnotation).factory();
        } catch (MirroredTypeException mte) {
            typeMirror = mte.getTypeMirror();
        }

        final String fullFactoryClassName = typeMirror.toString();
        final String className = fullFactoryClassName.substring(fullFactoryClassName.lastIndexOf(".") + 1);
        final String packageName = fullFactoryClassName.replace("." + className, "");

        ClassName interfaceClass = ClassName.get(packageName, className);
        TypeSpec.Builder implementationClass = TypeSpec.classBuilder(className + "Impl")
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addSuperinterface(interfaceClass);

        final String fullActivityClassName = typeElement.getQualifiedName().toString();
        implementMethod(implementationClass, fullActivityClassName, typeElement);

        return JavaFile.builder(packageName, implementationClass.build()).build();
    }

    private void implementMethod(TypeSpec.Builder builder, String fullActivityClassName, TypeElement typeElement) {
        for (Element innerElement : typeElement.getEnclosedElements()) {
            if (innerElement instanceof ExecutableElement &&
                innerElement.getKind() == ElementKind.METHOD &&
                innerElement.getAnnotation(SweetFactoryMethod.class) != null) {
                ExecutableElement executableElement = (ExecutableElement) innerElement;

                final StringBuilder parametersName = new StringBuilder();
                final String methodName = executableElement.getSimpleName().toString();
                MethodSpec.Builder methodSpec = MethodSpec.methodBuilder(methodName)
                        .addModifiers(Modifier.PUBLIC)
                        .addAnnotation(Override.class)
                        .returns(TypeName.get(executableElement.getReturnType()));

                final int parametersCount = executableElement.getParameters().size();
                for (int i = 0; i < parametersCount; i++) {
                    VariableElement parameter = executableElement.getParameters().get(i);

                    final String parameterName = parameter.getSimpleName().toString();

                    methodSpec.addParameter(TypeName.get(parameter.asType()), parameterName);
                    parametersName.append(parameterName);

                    if ((i + 1) < parametersCount) {
                        parametersName.append(", ");
                    }
                }

                final String returnStatement = String.format("return %s.%s(%s)", fullActivityClassName, methodName, parametersName.toString());
                methodSpec.addStatement(returnStatement);

                builder.addMethod(methodSpec.build());
            }
        }
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return new TreeSet<>(Collections.singletonList(
                SweetFactoryDeclaration.class.getCanonicalName()
        ));
    }

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }
}

File: sweetfactory-compiler/src/main/java/sweetfactory/compiler/SweetFactoryProcessor.kt
package sweetfactory.compiler

import com.google.auto.service.AutoService
import com.google.common.annotations.VisibleForTesting
import com.squareup.javapoet.JavaFile
import sweetfactory.annotations.SweetFactoryDeclaration
import java.io.IOException
import java.util.TreeSet
import javax.annotation.processing.Processor
import javax.annotation.processing.ProcessingEnvironment
import javax.annotation.processing.RoundEnvironment
import javax.annotation.processing.SupportedSourceVersion
import javax.annotation.processing.AbstractProcessor
import javax.lang.model.SourceVersion
import javax.lang.model.element.TypeElement
import kotlin.collections.HashMap

@AutoService(Processor::class)
@SupportedSourceVersion(SourceVersion.RELEASE_8)
class SweetFactoryProcessor : AbstractProcessor() {

    private lateinit var sweetFactoryDeclarationImplementer: SweetFactoryDeclarationImplementer

    override fun getSupportedAnnotationTypes(): Set<String> {
        return TreeSet(listOf(SweetFactoryDeclaration::class.java.canonicalName))
    }

    override fun getSupportedSourceVersion(): SourceVersion {
        return SourceVersion.latestSupported()
    }

    @VisibleForTesting
    internal fun init(processingEnvironment: ProcessingEnvironment?,
                      sweetFactoryDeclarationImplementer: SweetFactoryDeclarationImplementer) {
        super.init(processingEnvironment)
        this.sweetFactoryDeclarationImplementer = sweetFactoryDeclarationImplementer
    }

    override fun init(processingEnvironment: ProcessingEnvironment?) {
        this.init(processingEnvironment, SweetFactoryDeclarationImplementer())
    }

    override fun process(annotations: Set<TypeElement>, roundEnvironment: RoundEnvironment): Boolean {
        val factories = HashMap<String, FactoryImplementation>()

        roundEnvironment.getElementsAnnotatedWith(SweetFactoryDeclaration::class.java).forEach { element ->
            val typeElement = element as TypeElement

            val factoryClassName = sweetFactoryDeclarationImplementer.getClassNameOfFactory(typeElement)

            val factoryImplementation: FactoryImplementation
            if (factories.containsKey(factoryClassName)) {
                factoryImplementation = factories[factoryClassName]!!
            } else {
                factoryImplementation = FactoryImplementation.from(factoryClassName)
                factories[factoryClassName] = factoryImplementation
            }

            sweetFactoryDeclarationImplementer.implementMethods(factoryImplementation.typeSpecBuilder, typeElement)
        }

        factories.forEach { entry ->
            val javaFile = JavaFile.builder(
                entry.value.packageName,
                entry.value.typeSpecBuilder.build()
            ).build()

            try {
                javaFile.writeTo(processingEnv.filer)
            } catch (e: IOException) {
                val className = javaFile.toJavaFileObject().name.replace("/", ".")
                processingEnv.messager.error("Error when trying to generate code for %s", className)
            }
        }

        return false
    }

}
